src/processors/knex-processor.ts

```diff
@@ -164,7 +174,14 @@ export default class KnexProcessor<ResourceT extends Resource> extends Operation
     if (id) {
       return records[0];
     }
-    return records;
+
+    return new ResourcesOperationResult(
+      HttpStatusCode.OK,
+      records,
+      {
+        recordCount,
+      }
+    );
```

Comment:
@isBatak Processors are by definition transport-layer-agnostic. Unfortunately, this means that we cannot have a `HttpStatusCode.OK` here. Maybe if you can tell me a bit more about your reasoning here, we can come up with a different approach.

Reply:
I kinda agree with that rule, but how will `transport-layer` know which status to return back to the client. 
That information is owned by Processor.
We need some kind of abstraction between Processor and Transport Layer because there is multiple status codes that could be retuned by the Processor based on the settings, like on create https://jsonapi.org/format/1.1/#crud-creating-responses. 
But yeah Maybe I'm missing some detail here.
Main idea of `ResourcesOperationResult` is that it solves pagination problem, it holds the subset (one page) of data plus pagination metadata used for building links (LinkBuilder). Maybe I can focus on that for now...

Reply:
Kurier not only works with HTTP transports, but also with WebSockets. Because of this, processors can't use HTTP status codes: they aren't part of the response for WebSockets. Unless something goes wrong, jsonApiKoa/jsonApiExpress/etc. return 200, 201 or 204 (according to the HTTP method). If something goes wrong in a processor, we throw exceptions from the `JsonApiErrors` object. This pretty much solves status codes for us.l 

AFAIK, the spec doesn't require us to return data about the current page, the record count and other pagination info, only the links. In any case, such information could be returned in the `meta` section of the document.

Reply:
> Kurier not only works with HTTP transports, but also with WebSockets. Because of this, processors can't use HTTP status codes: they aren't part of the response for WebSockets. Unless something goes wrong, jsonApiKoa/jsonApiExpress/etc. return 200, 201 or 204 (according to the HTTP method). If something goes wrong in a processor, we throw exceptions from the JsonApiErrors object. This pretty much solves status codes for us.  

I understand, and I agree... I'll remove `HttpStatusCode.OK` from the `ResourcesOperationResult`.

> AFAIK, the spec doesn't require us to return data about the current page, the record count and other pagination info, only the links. In any case, such information could be returned in the meta section of the document. Introducing the `ResourcesOperationResult` type might have a significant impact across the codebase.

Idea and problem that `ResourcesOperationResult` solves is this, when we hit endpoint like `/book` operation result is `Array<Record>` so I don't have a place where I can store pagination info, except modifying Array props, which is weird. 
Also, it could store JsonApiParams so I don't have to prop drill them through the application methods (buildOperationResponse, serializeResources), and serializer to be able to build links.

Reply:
> Idea and problem that `ResourcesOperationResult` solves is this, when we hit endpoint like `/book` operation result is `Array<Record>` so I don't have a place where I can store pagination info, except modifying Array props, which is weird. 

OK, so this type would be a counterpart to `Operation`. So in a request/response model, we'd have `Operation` for the request` and `OperationResult` for the response. In `OperationResult` we'd need something to store something like `data` and `metadata` (`data` being the resources, `metadata` being anything extra, such as pagination info). Am I getting this right? If that's the case, I believe we can go ahead.

> Also, it could store JsonApiParams so I don't have to prop drill them through the application methods (buildOperationResponse, serializeResources), and serializer to be able to build links.

`JsonApiParams` is always available in the `Operation` instance. I'm guessing you _could_ store that as part of the response metadata, if that makes things easier to work with, but I'm not entirely sure if it's the best way to do it.